[23034] Improve settings caching

oliverguenther opened the pull request:
Currently, settings are cached on a per-name basis, but that cache is
expired everytime _any_ attribute is written.

Even when a week-old attribute is still valid, its cache is expired as
soon another setting is updated due to the max settings cache key.

Instead of caching a single attribute, I suggest to cache the entire
settings (the total number with plugins is less than 120 and 4KiB in
size) and retrieve and deserialize a single value from that hash.

That caching is two-fold, once using Rails.cache  for the entirety of the current settings hash (still potentially serialized to avoid additionally marshals).
For the current request, it also holds the settings in RequestStore.

That alleviates the previous effort:
1. Locating a single element from database
2. Instantiate a setting object just for deserializing
3. Requesting settings several times by using RequestStore

https://community.openproject.com/work_packages/23034/

/cc @machisuji , @crohr This is the result of what we discussed as part of the mail notifications PR.

oliverguenther committed:
Improve stored settings as a whole

Currently, settings are cached on a per-name basis, but that cache is
expired everytime *any* attribute is written.

Even when a week-old attribute is still valid, its cache is expired as
soon another setting is updated due to the max settings cache key.

Instead of caching a single attribute, I suggest to cache the entire
settings (the total number with plugins is less than 120 and 4KiB in
size) and retrieve and deserialize a single value from that hash.

That saves us the time to:

1. Locating a single element from database
2. Instantiate a setting object just for deserializing
3. Requesting settings several times by using RequestStore

oliverguenther committed:
Avoid caching on test env by default

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-fix-settings-caching.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-fix-settings-caching.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

oliverguenther commented on config/environments/test.rb:
```diff
@@ -69,7 +69,7 @@
   # config.action_view.raise_on_missing_translations = true
 
-  config.cache_store = :file_store, Rails.root.join("tmp", "cache", "paralleltests#{ENV['TEST_ENV_NUMBER']}")
+  config.cache_store = :null_store
```
This was introduced as part of the parallel testing feature. We never used _actual_ caching in tests before, and if we did _test_ caches, we used mocked as far as I can tell so we should rely on the null store instead.

oliverguenther commented on spec/spec_helper.rb:
```diff
@@ -92,6 +92,7 @@
 
     DatabaseCleaner.start
     ActionMailer::Base.deliveries.clear
+    RequestStore.clear!
```
`RequestStore` doesn't have any ties to the setting of `config.cache_store`, so it will store entries regardless of the test environment. Since it's a simple hash, I believe purging it before specs is the easiest / robust solution.

oliverguenther commented on app/models/setting.rb:
```diff
@@ -147,21 +142,30 @@ def value=(v)
 
   # Returns the value of the setting named name
   def self.[](name)
-    Marshal.load(Rails.cache.fetch(cache_key(name)) { Marshal.dump(find_or_default(name).value) })
```
We don't need to (un)serialize values here since `Rails.cache` does that already.

oliverguenther committed:
Rewrite with_setting specs

oliverguenther committed:
Set asset cache to memory

This enforces memory caching of assets in the test mode to avoid
breaking tests in parallel_test.

https://www.tkalin.com/blog_posts/rails-4-disable-assets-caching-for-development-test-environments/
